### PR TITLE
docs/DeveloperGuide: improve explanation for `find` feature.

### PR DESCRIPTION
**Pull request by lycjackie**
Let’s update the `DeveloperGuide` to improve our explanation to describe
the implementation better for our reader.

### CODE REVIEW

File: docs/DeveloperGuide.adoc [hunk 1]
```diff
@@ -375,7 +375,7 @@ the program.
 
 Step 1 is performed by the `ApplicationParser` class, and no special actions is needed for the `find` feature.
 
-Step 3 is performed by `BooleanExpression#parse`
+Step 2 and 3 is performed by `BooleanExpression#parse`
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: Should be `BooleanExpressionParser#parse(...)`

File: docs/DeveloperGuide.adoc [hunk 2]
```diff
@@ -390,10 +390,23 @@ image::FindCommandSequenceDiagram.png[width="650"]
 
 The main implementation of this feature is split into two components. The `Tokenizer` and `BooleanExpressionParser`
 
+`FindCommandParser` handles the users provided input and ensure that the provided is not empty. Once it passes the
+checks, the argument are passed to `BooleanExpressionParser` which handle the evaluation of the expression.
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: The mixed use of `input` and `argument` makes it confusing for readers.
I think you could simplify the second sentence by saying `FindCommandParser` calls `BooleanExpressionParser#parse(...)` to evaluate the expression.

File: docs/DeveloperGuide.adoc [hunk 3]
```diff
@@ -403,17 +416,50 @@ This class is initialized with the input argument and prefixes and can be querie
 Each query will consume the previous token and returns the next available token. +
 This is similar to how `java.util.Scanner` works.
 
+===== Design Consideration
+
+After many rounds of experiment with `StringTokenizer` provided by native java and `ArgumentMultimap`, we found 3 main
+issues which they do not provide.
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: Capitalise `Java`.
Also, you need to rephrase the second half of the sentences.

File: docs/DeveloperGuide.adoc [hunk 4]
```diff
@@ -403,17 +416,50 @@ This class is initialized with the input argument and prefixes and can be querie
...
+
+. `ArgumentMultimap` do not keep track of the order of each delimiter. They will only track if the delimiter provided
+exists.
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: `do` -> `does`

File: docs/DeveloperGuide.adoc (hunk 3 above)
Comment (Creastery) [reply docs/DeveloperGuide.adoc]: Here you mentioned 3 issues, but there are only two below!

File: docs/DeveloperGuide.adoc [hunk 5]
```diff
@@ -403,17 +416,50 @@ This class is initialized with the input argument and prefixes and can be querie
...
+
+. `StringTokenizer` default delimiter is whitespace, although we could change the delimiter and parse multiple
+delimiters. It does not suit our situation. +
+e.g. `find name/AAAA code/BBBB`. +
+This will return us 1 token. `name/AAAA code/BBBB`, which we could not use to check due to the missing `operator`. +
+We need the tokenizer to return us 2 tokens `name/AAAA` and `code/BBBB` in order for us to know that the expression
+was invalid due to the missing `operator`
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: Having worked on this with you on the `StringTokenizer`, I think this could be better phrased.
Discuss more about how we still need to do additional parsing after using `StringTokenizer` to facilitate the execution of Shunting Yard.

File: docs/DeveloperGuide.adoc [hunk 6]
```diff
@@ -403,17 +416,50 @@ This class is initialized with the input argument and prefixes and can be querie
...
+
+As such, our `Tokenizer` aim to solve these 2 issues.
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: `aim` -> `aims`.
Perhaps mentioned that we developed a custom `Tokenizer` that aims to solve these two issues.

File: docs/DeveloperGuide.adoc [hunk 7]
```diff
@@ -403,17 +416,50 @@ This class is initialized with the input argument and prefixes and can be querie
...
+
+The table below shows the different of our `Tokenizer` vs `StringTokenizer` and `ArgumentMultimap`
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: `different` -> `differences`.

File: docs/DeveloperGuide.adoc [hunk 8]
```diff
@@ -403,17 +416,50 @@ This class is initialized with the input argument and prefixes and can be querie
...
 This is represented by the class `pwe.planner.logic.parser.Operator` and define all valid operators to be used in
 `BooleanExpressionParser`.
 
+The table below shows the valid operators our application currently support.
+
+|====
+| *Operator* | *Description* | *Precedence*
+| `&&` | Logical "AND" operation (both conditions A AND B must match) | Highest
+| `\|\|` | Logical "OR" operation (either conditions A OR B must match)| Lowest
+|====
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc:493]: Since parenthesis is also considered as an operator, perhaps include it and talk about it briefly here.

File: docs/DeveloperGuide.adoc [hunk 9]
```diff
@@ -437,30 +483,75 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
 .Parser interactions
 image::parserSequenceDiagram.png[width="650"]
 
+When the `FindCommandParser` received the provided user argument, after doing it checks, it will pass the argument to
+ `BooleanExpressionParser` which will initialise a new tokenizer which extract all argument as token. +
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]:
```
initialise a new tokenizer which extract all argument as token
```
should be:
```
initialize a new `Tokenizer` that extracts the arguments as tokens
```

File: docs/DeveloperGuide.adoc [hunk 10]
```diff
@@ -437,30 +483,75 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+With each token, the `BooleanExpressionParser` will then create a predicate for each prefix.
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: This should be:
```
`BooleanExpressionParser` will create a `Predicate` based on the `Prefix` in each token. If the token is an `Operator`, `BooleanExpressionParser` will apply the operator on two `Predicate` to combine them into a composite `Predicate.

File: docs/DeveloperGuide.adoc [hunk 11]
```diff
@@ -437,30 +483,75 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+    throw new ParseException(
+            String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
+}
+return predicate;
+```
+
 ==== Design Consideration
 
-===== Aspect: How to parse composite predicates
+This sections share some design consideration we encounter when enhancing the existing `find` feature.
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: `share some design consideration we encounter` -> `details the design considerations we made`

File: docs/DeveloperGuide.adoc [hunk 12]
```diff
@@ -437,30 +483,75 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+
+===== Aspect: Parsing of composite predicate
 
 [NOTE]
 Currently we are using choice 2 because this provide our users with a more flexible way of finding modules in our
 application.
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: Might want to reorder the two points and remove this note.

File: docs/DeveloperGuide.adoc [hunk 13]
```diff
@@ -437,30 +483,75 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
-- Alternative 1: Do an implicit boolean `OR` for every predicate.
+- Approach 1: Do an implicit logical `OR` for every predicate.
 
-[cols="30%,<70%"]
+[cols="20%,<80%"]
 |======
 |*Pros*| Very easy to implementation
-|*Cons*| User will not be able to do a very detailed or complex find.
+|*Cons* a|
+* Returns a very huge set of results when the keywords increase.
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: `keywords increase` -> `number of keywords increase`.

File: docs/DeveloperGuide.adoc [hunk 14]
```diff
@@ -437,30 +483,75 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+* Find command will be restricted to logical `OR` results.
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc:586]: Could be better phrased.

File: docs/DeveloperGuide.adoc [hunk 15]
```diff
@@ -437,30 +483,75 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
 |======
 
-- Alternative 2 (current choice): Implement an algorithm https://en.wikipedia.org/wiki/Shunting-yard_algorithm[Shunting Yard]
- that can parse complex boolean expression.
+- Approach 2 (current choice): Implement an algorithm https://en.wikipedia.org/wiki/Shunting-yard_algorithm[Shunting
+Yard]
+ that can parse complex boolean expression and composite the predicate.
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: You can't just `composite the predicate`. `Composite` isn't an action verb!

File: docs/DeveloperGuide.adoc [hunk 16]
```diff
@@ -437,30 +483,75 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
 
-[cols="30%,<70%"]
+[cols="20%,<80%"]
 |=====
-| *Pros* | User will have very flexible searching terms
-| *Cons* | Developer will take a very long time to implement and test for edge cases.
+| *Pros* | Find command can be very flexible as per user wish.
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: Could go into a bit more details here.

File: docs/DeveloperGuide.adoc [hunk 17]
```diff
@@ -437,30 +483,75 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+| *Cons* a|
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: Remove `a`.

File: docs/DeveloperGuide.adoc [hunk 18]
```diff
@@ -437,30 +483,75 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+* Requires a lot of tests to ensure expected behaviour
+* Extra work required to integrate a new logic parser into `FindCommandParser` and may cause build problems if not
+implemented properly.
 |=====
+
+After weighing both approach pros and cons, we have decided to go with the *approach 2*. +
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: `we have decided` -> `we decided`.

File: docs/DeveloperGuide.adoc [hunk 19]
```diff
@@ -437,30 +483,75 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+This is because we are expecting many common names between a module available in the university curriculum. If the
+user could only search with an implicit logical `OR`, they would not be able to find a module which suit them
+effectively. Which defeat the purpose of our application.
```
Comment (Creastery) [inline docs/DeveloperGuide.adoc]: `Which defeat the purpose of our application` is a strong statement.
Perhaps you could say: 
```
drastically reduce the effectiveness of the `find` command
```

**Review by Creastery — CHANGES_REQUESTED** [review]

File: docs/DeveloperGuide.adoc (hunk 17 above)
Comment (lycjackie) [reply docs/DeveloperGuide.adoc]: This is intended.
 
```
a| 
* point 1
* point 2
```

Allows the cells to have bullets point.

File: docs/DeveloperGuide.adoc [hunk 20]
```diff
@@ -362,7 +362,7 @@ refreshed to display the full listing just to be able to iterate and delete modu
 
 The find feature aims to help users to be able to locate any module in our application easily. We support
```
Comment (truegitnovice) [inline docs/DeveloperGuide.adoc]: I think it will be better to change this part too. 
```suggestion
The find feature aims to help users to be able to easily locate any module in our application. 
```

Comment (Letm3through) [inline docs/DeveloperGuide.adoc]: I think it will be better to change this part too. 
```suggestion
The find feature aims to help users to be able to easily locate any modules in our application. 
```

File: docs/DeveloperGuide.adoc [hunk 21]
```diff
@@ -362,7 +362,7 @@ refreshed to display the full listing just to be able to iterate and delete modu
...
 the finding of module's name, code and credits. This enable our users to be able to find for any modules with partial
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
We support the finding of modules with their names, codes and credits. This enables our users to find any modules with partial
```

File: docs/DeveloperGuide.adoc [hunk 22]
```diff
@@ -362,7 +362,7 @@ refreshed to display the full listing just to be able to iterate and delete modu
...
- information.
+ knowledge about the module they need.
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]: Do you mean this?
```suggestion
information of the modules they have.
```

File: docs/DeveloperGuide.adoc [hunk 23]
```diff
@@ -390,10 +390,22 @@ image::FindCommandSequenceDiagram.png[width="650"]
 
 The main implementation of this feature is split into two components. The `Tokenizer` and `BooleanExpressionParser`
 
+`FindCommandParser` calls `BooleanExpressionParser#parse` which handle the evaluation of the expression.
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
`FindCommandParser` calls `BooleanExpressionParser#parse` which handles the evaluation of the expression.
```

File: docs/DeveloperGuide.adoc [hunk 24]
```diff
@@ -390,10 +390,22 @@ image::FindCommandSequenceDiagram.png[width="650"]
...
+
 1. `Tokenizer` helps to split the user provided argument into tokens which could be used by `BooleanExpressionParser`.
 2. `BooleanExpressionParser` handle the high-level syntax of the find expressions and map them into
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
2. `BooleanExpressionParser` handles the high-level syntax of the find expressions and maps them into
```

File: docs/DeveloperGuide.adoc [hunk 25]
```diff
@@ -403,17 +415,50 @@ This class is initialized with the input argument and prefixes and can be querie
 Each query will consume the previous token and returns the next available token. +
 This is similar to how `java.util.Scanner` works.
 
+===== Design Consideration
+
+After many rounds of experiment with `StringTokenizer` provided by native Java and `ArgumentMultimap`. +
+We found two main issues which could not be satisfy by either `StringTokenizer` or `ArgumentMultimap`.
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
We found two main issues which could not be satisfied by either `StringTokenizer` or `ArgumentMultimap`.
```

File: docs/DeveloperGuide.adoc [hunk 26]
```diff
@@ -403,17 +415,50 @@ This class is initialized with the input argument and prefixes and can be querie
...
+
+. `ArgumentMultimap` does not keep track of the order of each delimiter. They will only track if the delimiter provided
+exists.
+
+. `StringTokenizer` default delimiter is whitespace, although we could change the delimiter and parse multiple
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
. `StringTokenizer` has default delimiter as whitespace, although we could change the delimiter and parse multiple
```

File: docs/DeveloperGuide.adoc [hunk 27]
```diff
@@ -403,17 +415,50 @@ This class is initialized with the input argument and prefixes and can be querie
...
+|====
+| *`Tokenizer`* | *`StringTokenizer`* | *`ArgumentMultimap`*
+| Respect the order of delimiters. | Respect the order of delimiters | Only keep track which delimiters is present.
+| Support multiple delimiters. | Support only multiple delimiters. | Support multiple delimiters.
+| More flexible in splitting of token | Only split token based on delimiters | Does not split into token.
+|====
+
 ==== Operator
 
 This is represented by the class `pwe.planner.logic.parser.Operator` and define all valid operators to be used in
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
This is represented by the class `pwe.planner.logic.parser.Operator` and defines all valid operators to be used in
```

File: docs/DeveloperGuide.adoc [hunk 28]
```diff
@@ -437,30 +482,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
 .Parser interactions
 image::parserSequenceDiagram.png[width="650"]
 
-==== Design Consideration
+When the `FindCommandParser` received the provided user argument, after doing it checks, it will pass the argument to
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]: Maybe you may want to elaborate more on what _checks_ is carried out?
```suggestion
When `FindCommandParser` receives the provided user argument, it will carry out checks and pass the argument to
```

File: docs/DeveloperGuide.adoc [hunk 29]
```diff
@@ -437,30 +482,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+}
+return predicate;
+```
 
-[cols="30%,<70%"]
-|======
-|*Pros*| Very easy to implementation
-|*Cons*| User will not be able to do a very detailed or complex find.
-|======
+==== Design Consideration
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
==== Design Considerations
```

File: docs/DeveloperGuide.adoc [hunk 30]
```diff
@@ -437,30 +482,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
 
-- Alternative 2 (current choice): Implement an algorithm https://en.wikipedia.org/wiki/Shunting-yard_algorithm[Shunting Yard]
- that can parse complex boolean expression.
+This sections share details the design considerations we made when enhancing the existing `find` feature.
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
This section shares the design considerations we went through during the enhancement of the existing `find` feature.
```

File: docs/DeveloperGuide.adoc [hunk 31]
```diff
@@ -437,30 +482,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
-[cols="30%,<70%"]
+===== Aspect: Parsing of composite predicate
+The table below shows the comparison of different approach we could take.
+
+[cols="30%,35%,35%"]
 |=====
-| *Pros* | User will have very flexible searching terms
-| *Cons* | Developer will take a very long time to implement and test for edge cases.
+|*Approach* | *Pros* | *Cons*|
+*1.* Implement an algorithm https://en.wikipedia.org/wiki/Shunting-yard_algorithm[Shunting Yard] that parse the complex
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
*1.* Implement an algorithm https://en.wikipedia.org/wiki/Shunting-yard_algorithm[Shunting Yard] that parses the complex
```

File: docs/DeveloperGuide.adoc [hunk 32]
```diff
@@ -437,30 +482,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
-|======
+==== Design Consideration
 
-- Alternative 2 (current choice): Implement an algorithm https://en.wikipedia.org/wiki/Shunting-yard_algorithm[Shunting Yard]
- that can parse complex boolean expression.
+This sections share details the design considerations we made when enhancing the existing `find` feature.
 
-[cols="30%,<70%"]
+===== Aspect: Parsing of composite predicate
+The table below shows the comparison of different approach we could take.
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
The table below shows comparisons between the two implementations that could have been made.
```

File: docs/DeveloperGuide.adoc [hunk 33]
```diff
@@ -437,30 +482,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+[cols="30%,35%,35%"]
 |=====
-| *Pros* | User will have very flexible searching terms
-| *Cons* | Developer will take a very long time to implement and test for edge cases.
+|*Approach* | *Pros* | *Cons*|
+*1.* Implement an algorithm https://en.wikipedia.org/wiki/Shunting-yard_algorithm[Shunting Yard] that parse the complex
+boolean expression and returns a composite predicate. |
+Find command can be very flexible and scale with the application data.
+a|
+* Requires a lot of tests to ensure expected behaviour
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
* Requires many tests to ensure expected behaviours
```

File: docs/DeveloperGuide.adoc [hunk 34]
```diff
@@ -437,30 +482,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+* Extra work required to integrate a new logic parser into `FindCommandParser` and may cause build problems if not
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
* Extra work is required to integrate a new logic parser into `FindCommandParser`. It may result in build problems if it is not
```

File: docs/DeveloperGuide.adoc [hunk 35]
```diff
@@ -437,30 +482,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+The table below shows the comparison of different approach we could take.
+
+[cols="30%,35%,35%"]
 |=====
-| *Pros* | User will have very flexible searching terms
-| *Cons* | Developer will take a very long time to implement and test for edge cases.
+|*Approach* | *Pros* | *Cons*|
+*1.* Implement an algorithm https://en.wikipedia.org/wiki/Shunting-yard_algorithm[Shunting Yard] that parse the complex
+boolean expression and returns a composite predicate. |
+Find command can be very flexible and scale with the application data.
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]: Sounds quite vague.
```suggestion
Find command can be very flexible. It can work with many arguments to search for the specific modules that the user wants.
```

File: docs/DeveloperGuide.adoc [hunk 36]
```diff
@@ -437,30 +482,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+* Requires a lot of tests to ensure expected behaviour
+* Extra work required to integrate a new logic parser into `FindCommandParser` and may cause build problems if not
+implemented properly. |
+*2.* Do an implicit logical `OR` for every predicate. |
+Very easy to implement a|
+* Returns a very huge set of results when the number of keywords increase.
+* Find command will be restricted to logical `OR` results.
 |=====
+
+After weighing both approach pros and cons, we decided to go with the *approach 1*. +
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
After weighing both pros and cons, we decided to go with the *approach 1*. +
```

File: docs/DeveloperGuide.adoc [hunk 37]
```diff
@@ -437,30 +482,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+This is because we are expecting many common names between a module available in the university curriculum. If the
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
This is because we are expecting many similar names between modules available in the university curriculum. If the
```

File: docs/DeveloperGuide.adoc [hunk 38]
```diff
@@ -437,30 +482,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+user could only search with an implicit logical `OR`, they would not be able to find a module which suit them
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
user could only search with an implicit logical `OR`, the user would not be able to find a module which suits the user's needs
```

File: docs/DeveloperGuide.adoc [hunk 39]
```diff
@@ -437,30 +482,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+effectively. Which drastically reduce the effectiveness of the `find` command.
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
effectively. This can drastically reduce the effectiveness of the `find` command.
```

**Review by Letm3through — CHANGES_REQUESTED** [review]

File: docs/DeveloperGuide.adoc [hunk 40]
```diff
@@ -403,17 +415,55 @@ This class is initialized with the input argument and prefixes and can be querie
...
+. `StringTokenizer` can take in `PREFIX` as delimiter, however, this will split the argument (i.e. `code/CS1231`) into
+two tokens, `code/` and `CS1231`. Additional parsing is required before we are able to pass it to
+`BooleanExpressionParser`. These parsing creates more overhead as we have to ensure there are proper checks is done.
+
+As such, our custom `Tokenizer` aims to solve these three issues.
+
+The table below shows the differences of our `Tokenizer` vs `StringTokenizer` and `ArgumentMultimap`
+|====
+| *`Tokenizer`* | *`StringTokenizer`* | *`ArgumentMultimap`*
+| Respect the order of delimiters. | Respect the order of delimiters | Only keep track which delimiters is present.
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
| Respect the order of delimiters. | Respect the order of delimiters | Only keep track which delimiters are present.
```

File: docs/DeveloperGuide.adoc [hunk 41]
```diff
@@ -437,30 +487,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+* Requires many tests to ensure expected behaviours
+* Extra work is required to integrate a new logic parser into `FindCommandParser`. It may result in  build problems if
+ it is not implemented properly.
+ |*2.* Do an implicit logical `OR` for every predicate.
+ |Very easy to implement a|
+* Returns a very huge set of results when the number of keywords increase.
+* Find command will be restricted to logical `OR` results.
 |=====
+
+After weighing both pros and cons, we decided to go with the *approach 1*. +
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
After weighing both pros and cons, we decided to go with *approach 1*. +
```

File: docs/DeveloperGuide.adoc [hunk 42]
```diff
@@ -403,17 +415,55 @@ This class is initialized with the input argument and prefixes and can be querie
 Each query will consume the previous token and returns the next available token. +
 This is similar to how `java.util.Scanner` works.
 
+===== Design Consideration
+
+After many rounds of experiment with `StringTokenizer` provided by native Java and `ArgumentMultimap`. +
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
After many rounds of experiment with `StringTokenizer` that is provided by native Java and `ArgumentMultimap`. +
```

File: docs/DeveloperGuide.adoc [hunk 43]
```diff
@@ -403,17 +415,55 @@ This class is initialized with the input argument and prefixes and can be querie
...
+We found three main issues which could not be satisfied by either `StringTokenizer` or `ArgumentMultimap`.
+
+. `ArgumentMultimap` does not keep track of the order of each delimiter. They will only track if the delimiter provided
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
. `ArgumentMultimap` does not keep track of the order of each delimiter. They will only track if the delimiter is provided
```

File: docs/DeveloperGuide.adoc [hunk 44]
```diff
@@ -403,17 +415,55 @@ This class is initialized with the input argument and prefixes and can be querie
...
+exists.
+
+. `StringTokenizer` has default delimiter as whitespace, although we could change the delimiter and parse multiple
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
. `StringTokenizer` has a default delimiter as a whitespace, although we could change the delimiter and parse multiple
```

File: docs/DeveloperGuide.adoc [hunk 45]
```diff
@@ -403,17 +415,55 @@ This class is initialized with the input argument and prefixes and can be querie
...
+delimiters. It does not suit our situation. +
+e.g. `find name/AAAA code/BBBB`. +
+This will return us 1 token. `name/AAAA code/BBBB`, which we could not use to check due to the missing `operator`. +
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
This will return us 1 token. `name/AAAA code/BBBB`, which we cannot use to check due to the missing `operator`. +
```
`cannot` since I believe it implies now too

File: docs/DeveloperGuide.adoc [hunk 46]
```diff
@@ -403,17 +415,55 @@ This class is initialized with the input argument and prefixes and can be querie
...
+We need the tokenizer to return us 2 tokens `name/AAAA` and `code/BBBB` in order for us to know that the expression
+was invalid due to the missing `operator`.
+. `StringTokenizer` can take in `PREFIX` as delimiter, however, this will split the argument (i.e. `code/CS1231`) into
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc:465]: You may want to add a newline after line 431 just like you did at other lines.

File: docs/DeveloperGuide.adoc [hunk 47]
```diff
@@ -403,17 +415,55 @@ This class is initialized with the input argument and prefixes and can be querie
...
+two tokens, `code/` and `CS1231`. Additional parsing is required before we are able to pass it to
+`BooleanExpressionParser`. These parsing creates more overhead as we have to ensure there are proper checks is done.
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
`BooleanExpressionParser`. This parsing creates more overhead as we have to ensure that proper checks are carried out.
```

File: docs/DeveloperGuide.adoc [hunk 48]
```diff
@@ -403,17 +415,55 @@ This class is initialized with the input argument and prefixes and can be querie
...
+
+As such, our custom `Tokenizer` aims to solve these three issues.
+
+The table below shows the differences of our `Tokenizer` vs `StringTokenizer` and `ArgumentMultimap`
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
The table below shows the differences among our `Tokenizer`, `StringTokenizer` and `ArgumentMultimap`
```

File: docs/DeveloperGuide.adoc [hunk 49]
```diff
@@ -403,17 +415,55 @@ This class is initialized with the input argument and prefixes and can be querie
...
+|====
+| *`Tokenizer`* | *`StringTokenizer`* | *`ArgumentMultimap`*
+| Respect the order of delimiters. | Respect the order of delimiters | Only keep track which delimiters is present.
+| Support multiple delimiters. | Support only multiple delimiters. | Support multiple delimiters.
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
| Support multiple delimiters. | Support multiple delimiters. | Support multiple delimiters.
```
Is there any reason why is there `only` in just that one?

File: docs/DeveloperGuide.adoc [hunk 50]
```diff
@@ -403,17 +415,55 @@ This class is initialized with the input argument and prefixes and can be querie
...
+| More flexible in splitting of token | Only split token based on delimiters | Does not split into token.
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
| Splitting tokens is more flexible | Only split token based on delimiters | Does not split into token.
```

File: docs/DeveloperGuide.adoc [hunk 51]
```diff
@@ -403,17 +415,55 @@ This class is initialized with the input argument and prefixes and can be querie
...
+|====
+
 ==== Operator
 
-This is represented by the class `pwe.planner.logic.parser.Operator` and define all valid operators to be used in
+This is represented by the class `pwe.planner.logic.parser.Operator` and defines all valid operators to be used in
 `BooleanExpressionParser`.
 
+The table below shows the valid operators our application currently support.
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
The table below shows the valid operators that our application currently support.
```

File: docs/DeveloperGuide.adoc [hunk 52]
```diff
@@ -437,30 +487,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
-[cols="30%,<70%"]
-|======
-|*Pros*| Very easy to implementation
-|*Cons*| User will not be able to do a very detailed or complex find.
-|======
+==== Design Considerations
 
-- Alternative 2 (current choice): Implement an algorithm https://en.wikipedia.org/wiki/Shunting-yard_algorithm[Shunting Yard]
- that can parse complex boolean expression.
+This sections shares the design considerations we went through during the enhancing the existing `find` feature.
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
This section shares the design considerations we went through during the enhancement of the existing `find` feature.
```

File: docs/DeveloperGuide.adoc [hunk 53]
```diff
@@ -437,30 +487,66 @@ The sequence diagram below shows the interactions between `FindCommandParser` an
...
+ it is not implemented properly.
+ |*2.* Do an implicit logical `OR` for every predicate.
+ |Very easy to implement a|
+* Returns a very huge set of results when the number of keywords increase.
+* Find command will be restricted to logical `OR` results.
 |=====
+
+After weighing both pros and cons, we decided to go with the *approach 1*. +
+This is because we are expecting many similar names between modules in the university curriculum. If the
+user could only search with an implicit logical `OR`, the user would not be able to find a module which suit the
```
Comment (Letm3through) [inline docs/DeveloperGuide.adoc]:
```suggestion
user could only search with an implicit logical `OR`, the user would not be able to find modules which suit the
```

**Review by Letm3through — CHANGES_REQUESTED** [review]

**Review by Creastery — APPROVED** [review]
LGTM. Approving and merging.

**Review by Letm3through — APPROVED** [review]
